sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -73,132 +79,167 @@
  * <p>The source also accepts an optional configuration: {@code withQueryFilter()} allows you to
  * define a JSON filter to get subset of files in the database.</p>
  *
- * <p>There is also an optional {@code ParseCallback} that can be specified that can be used to
+ * <p>There is also an optional {@code Parser} that can be specified that can be used to
  * parse the InputStream into objects usable with Beam.  By default, MongoDbGridFSIO will parse
  * into Strings, splitting on line breaks and using the uploadDate of the file as the timestamp.
+ * When using a parser that outputs with custom timestamps, you may also need to specify
+ * the allowedTimestampSkew option.
  */
 public class MongoDbGridFSIO {
 
   /**
-   * Function for parsing the GridFSDBFile into objects for the PCollection.
+   * Callback for the parser to use to submit data.
+   */
+  public interface ParserCallback<T> extends Serializable {
+    public void output(T output);
+
+    public void output(T output, @Nullable Instant timestamp);
```

Comment:
What do you envision users would specify as timestamps instead of the default?
E.g. does this provide value over applying WithTimestamps transform over the output of this transform?